include/thread_pool/rouser.hpp

```diff
@@ -0,0 +1,160 @@
+#include <thread>
+#include <limits>
+#include <mutex>
+#include <chrono>
+#include <condition_variable>
+
+namespace tp
+{
+
+/**
+* @brief Rouser is a specialized worker that periodically wakes other workers. 
+* @detail This serves two purposes:
+* The first is that it emplaces an upper bound on the processing time of tasks in the thread pool, since
+* it is otherwise possible for the thread pool to enter a state where all threads are asleep, and tasks exist
+* in worker queues. The second is that it increases the likelihood of at least one worker busy-waiting at any
+* point in time, which speeds up task processing response time.
+*/
+class Rouser final
+{
+public:
+    /**
+    * @brief Worker Constructor.
+    */
+    Rouser(std::chrono::microseconds rouse_period);
+
+    /**
+    * @brief Copy ctor implementation.
+    */
+    Rouser(Rouser const&) = delete;
+
+    /**
+    * @brief Copy assignment implementation.
+    */
+    Rouser& operator=(Rouser const& rhs) = delete;
+
+    /**
+    * @brief Move ctor implementation.
+    * @note Be very careful when invoking this while the thread pool is 
+    * active, or in an otherwise undefined state.
+    */
+    Rouser(Rouser&& rhs) noexcept;
+
+    /**
+    * @brief Move assignment implementaion.
+    * @note Be very careful when invoking this while the thread pool is 
+    * active, or in an otherwise undefined state.
+    */
+    Rouser& operator=(Rouser&& rhs) noexcept;
+
+    /**
+    * @brief Destructor implementation.
+    */
+    ~Rouser();
+
+    /**
+    * @brief start Create the executing thread and start tasks execution.
+    * @param workers A pointer to the vector containing sibling workers for performing round robin work stealing.
+    * @param idle_workers A pointer to the slotted bag containing all idle workers.
+    * @param num_busy_waiters A pointer to the atomic busy waiter counter.
+    * @note The parameters passed into this function generally relate to the global thread pool state.
+    */
+    template <typename Task, template<typename> class Queue>
+    void start(std::vector<std::unique_ptr<Worker<Task, Queue>>>& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters);
+
+    /**
+    * @brief stop Stop all worker's thread and stealing activity.
+    * Waits until the executing thread becomes finished.
+    */
+    void stop();
+
+private:
+
+    /**
+    * @brief threadFunc Executing thread function.
+    * @param workers A pointer to the vector containing sibling workers for performing round robin work stealing.
+    * @param idle_workers A pointer to the slotted bag containing all idle workers.
+    * @param num_busy_waiters A pointer to the atomic busy waiter counter.
+    */
+    template <typename Task, template<typename> class Queue>
+    void threadFunc(std::vector<std::unique_ptr<Worker<Task, Queue>>>& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters);
+
+    std::atomic<bool> m_running_flag;
```

Comment:
Just a general recommendation with designing this type of logic controlled with flags. You could alternatively have a single atomic variable which represents the states of the object. When you have one variable representing the state it's easier to control transitions between the states with exchange and compare exchange operations and easier to reason about the state.

For this class you can model the following states:
not started
start requested
running
stop requested
stopped

This is not a necessary change now, but if you have time later you can experiment with it.

Reply:
My reasoning for keeping the states separate was that the `m_running_flag` check in the `threadFunc` loop would be faster with boolean comparison versus an enum. 

Not sure if the performance gain is worth the hit to readability?

Reply:
I concur with Hayk. Modeling this with an enum rather than a bunch of boolean flags would make your life much easier. The difference in performance should be so small as to be negligible, AFAIK.

Reply:
👍 will change.